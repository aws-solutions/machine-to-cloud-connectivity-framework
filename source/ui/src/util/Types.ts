// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import React from 'react';

export type ErrorMessage = { errorMessage: string };
export type FormControlElement = HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement;
export type KeyStringValue = Record<string, string | undefined>;
type ApiMethod = 'get' | 'post';
type GreengrassCoreDeviceStatus = 'HEALTHY' | 'UNHEALTHY';
type LoggingLevel = 'ERROR' | 'WARN' | 'INFO' | 'DEBUG' | 'VERBOSE';

export enum ConnectionControl {
  DEPLOY = 'deploy',
  START = 'start',
  STOP = 'stop',
  UPDATE = 'update',
  DELETE = 'delete',
  PUSH = 'push',
  PULL = 'pull',
  FAIL = 'fail'
}

export enum LogType {
  INFO = 'info',
  ERROR = 'error'
}

export enum MachineProtocol {
  OPCDA = 'opcda',
  OPCUA = 'opcua',
  OSIPI = 'osipi',
  MODBUSTCP = 'modbustcp'
}

export enum MessageModalType {
  MESSAGE,
  CONFIRM
}

export enum PaginationType {
  PREV,
  NEXT
}

export enum CreatedBy {
  SYSTEM = 'System',
  USER = 'User'
}

export enum OsPlatform {
  LINUX = 'linux',
  WINDOWS = 'windows'
}

export enum GreengrassCoreDeviceControl {
  CREATE = 'create',
  DELETE = 'delete'
}

export enum OsiPiAuthMode {
  ANONYMOUS = 'ANONYMOUS',
  BASIC = 'BASIC'
  //TODO: Figure out how to support Kerberos auth
  // KERBEROS = 'KERBEROS'
}

export interface AmplifyConfigurationInput {
  apiEndpoint: string;
  identityPoolId: string;
  loggingLevel: LoggingLevel;
  region: string;
  s3Bucket: string;
  userPoolId: string;
  webClientId: string;
}

export interface AmplifyConfigurationResponse {
  API: {
    endpoints: { name: string; endpoint: string; region: string }[];
  };
  Auth: {
    identityPoolId: string;
    region: string;
    userPoolId: string;
    userPoolWebClientId: string;
  };
  Storage: {
    AWSS3: {
      bucket: string;
      region: string;
    };
  };
}

export interface ConnectionDefinition {
  control: ConnectionControl;
  connectionName: string;
  protocol: MachineProtocol;
  area?: string;
  greengrassCoreDeviceName?: string;
  machineName?: string;
  opcDa?: OpcDaDefinition;
  opcUa?: OpcUaDefinition;
  osiPi?: OsiPiDefinition;
  modbusTcp?: ModbusTcpDefinition;
  process?: string;
  sendDataToIoTSiteWise?: boolean;
  sendDataToIoTTopic?: boolean;
  sendDataToKinesisDataStreams?: boolean;
  sendDataToTimestream?: boolean;
  sendDataToHistorian?: boolean;
  historianKinesisDatastreamName?: string;
  siteName?: string;
  logLevel?: string;
}

export interface CommonDefinition {
  machineIp: string;
  serverName: string;
}

export interface OpcDaDefinition extends CommonDefinition {
  interval: number | string;
  iterations: number | string;
  listTags?: string[];
  tags?: string[];
}

export interface OpcUaDefinition extends CommonDefinition {
  port?: number | string;
}

export interface OsiPiDefinition {
  apiUrl: string;
  serverName: string;
  authMode: OsiPiAuthMode;
  verifySSL: boolean;
  username?: string;
  password?: string;
  tags: string[];
  requestFrequency: number | string;
  catchupFrequency: number | string;
  maxRequestDuration: number | string;
  queryOffset: number | string;
}

export interface ModbusTcpDefinition {
  host: string;
  hostPort: number | string;
  hostTag: string;
  modbusSlavesConfigSerialized: string;
  modbusSlavesConfig: ModbusTcpSlaveDefinition[];
}

export interface ModbusTcpSlaveDefinition {
  slaveAddress: number | string;
  frequencyInSeconds: number;
  commandConfig: ModbusTcpSlaveCommandConfig;
}

export interface ModbusTcpSlaveCommandConfig {
  readCoils?: ModbusTcpSlaveCommandReadCoilsConfig;
  readDiscreteInputs?: ModbusTcpSlaveCommandReadDiscreteInputsConfig;
  readHoldingRegisters?: ModbusTcpSlaveCommandReadHoldingRegistersConfig;
  readInputRegisters?: ModbusTcpSlaveCommandReadInputRegistersConfig;
}

export interface ModbusTcpSlaveCommandIndividualConfig {
  address: number | string;
  count?: number | string;
}

export type ModbusTcpSlaveCommandReadCoilsConfig = ModbusTcpSlaveCommandIndividualConfig;
export type ModbusTcpSlaveCommandReadDiscreteInputsConfig = ModbusTcpSlaveCommandIndividualConfig;
export type ModbusTcpSlaveCommandReadHoldingRegistersConfig = ModbusTcpSlaveCommandIndividualConfig;
export type ModbusTcpSlaveCommandReadInputRegistersConfig = ModbusTcpSlaveCommandIndividualConfig;

export interface ListConnectionsItem {
  connectionName: string;
  machineName: string;
  protocol: MachineProtocol;
  status: ConnectionControl;
  sendDataToIoTSiteWise: boolean;
  sendDataToIoTTopic: boolean;
  sendDataToKinesisDataStreams: boolean;
  sendDataToHistorian: boolean;
}

export interface ListConnectionsResponse {
  connections: ListConnectionsItem[];
  nextToken?: string;
}

export interface ControlConnectionResponse {
  connectionName: string;
  control: ConnectionControl;
  result: string;
}

export interface CreateUpdateConnectionResponse {
  connectionName: string;
  control: ConnectionControl.DEPLOY | ConnectionControl.UPDATE;
  message: string;
}

export interface ModalProps {
  show: boolean;
  hide: () => void;
}

export interface ConnectionLogsModalProps extends ModalProps {
  connectionName: string;
}

export interface MessageModalProps extends ModalProps {
  message: string | React.ReactNode;
  modalType: MessageModalType;
  confirmAction?: () => void | Promise<void>;
}

export interface GetConnectionResponse {
  connectionName: string;
  greengrassCoreDeviceName: string;
  protocol: MachineProtocol;
  sendDataToIoTSiteWise: boolean;
  sendDataToIoTTopic: boolean;
  sendDataToKinesisDataStreams: boolean;
  sendDataToTimestream: boolean;
  sendDataToHistorian: boolean;
  area?: string;
  machineName?: string;
  logLevel?: string;
  opcDa?: OpcDaDefinition;
  opcUa?: OpcUaDefinition;
  osiPi?: OsiPiDefinition;
  modbusTcp?: ModbusTcpDefinition;
  process?: string;
  siteName?: string;
  opcDaListTags?: string;
  opcDaTags?: string;
  osiPiTags?: string;
  historianKinesisDatastreamName?: string;
}

export interface FormProps {
  connection: GetConnectionResponse;
  onChange: (event: React.ChangeEvent<FormControlElement>) => void;
  errors: KeyStringValue;
}

export interface OpcUaFormProps extends FormProps {
  region: string;
}

export interface ListLogsItem {
  connectionName: string;
  timestamp: number;
  logType: LogType;
  message: string;
}

export interface ListLogsResponse {
  logs: ListLogsItem[];
  nextToken?: string;
}

export interface AlphaNumericValidationProps {
  value: string;
  maxLength: number;
  errors: KeyStringValue;
  errorKeyName: string;
  errorMessage: string;
}

export interface ErrorProps {
  message?: string;
  response?: {
    data: ErrorMessage | unknown;
  };
}

export interface ApiProps {
  method: ApiMethod;
  path: string;
  options?: Record<string, unknown>;
}

export interface ListGreengrassCoreDevicesResponse {
  greengrassCoreDevices: ListGreengrassCoreDevicesItem[];
  nextToken?: string;
}

export interface ListGreengrassCoreDevicesItem {
  name: string;
  createdBy: CreatedBy;
  numberOfConnections: number;
  osPlatform: string;
}

export interface UserGreengrassCoreDeviceResponse {
  greengrassCoreDevices: UserGreengrassCoreDevice[];
}

export interface UserGreengrassCoreDevice {
  coreDeviceThingName: string;
  status: GreengrassCoreDeviceStatus;
}

export interface GreengrassCoreDevicePostResponse {
  name: string;
  message: string;
}

export interface GreengrassCoreDeviceDomIds {
  category: string;
  greengrassCoreDeviceName: string;
  osPlatform: string;
}
