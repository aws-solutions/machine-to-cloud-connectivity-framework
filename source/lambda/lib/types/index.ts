// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export * as ConnectionBuilderTypes from './connection-builder-types';
export * as CustomResourceTypes from './custom-resource-types';
export * as DynamoDBHandlerTypes from './dynamodb-handler-types';
export * as GreengrassDeployerTypes from './greengrass-deployer-types';
export * as GreengrassV2HandlerTypes from './greengrass-v2-handler-types';
export * as IoTHandlerTypes from './iot-handler-types';
export * as IoTSiteWiseHandlerTypes from './iot-sitewise-handler-types';
export * as MessageConsumerTypes from './message-consumer-types';
export * as S3HandlerTypes from './s3-handler-types';
export * as SolutionCommonTypes from './solution-common-types';
export * as UtilsTypes from './utils-types';
