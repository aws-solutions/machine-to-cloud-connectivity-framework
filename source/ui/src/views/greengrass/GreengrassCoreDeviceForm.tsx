// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { I18n, Logger } from '@aws-amplify/core';
import React, { FormEvent, useState } from 'react';
import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import Breadcrumb from 'react-bootstrap/Breadcrumb';
import Card from 'react-bootstrap/Card';
import Col from 'react-bootstrap/Col';
import Container from 'react-bootstrap/Container';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';
import { LinkContainer } from 'react-router-bootstrap';
import GreengrassCoreDeviceInputForm from './GreengrassCoreDeviceInputForm';
import { handleValueChange } from './greengrass-core-device-form-utils';
import MessageModal from '../../components/MessageModal';
import EmptyCol from '../../components/EmptyCol';
import EmptyRow from '../../components/EmptyRow';
import { LoadingSpinner } from '../../components/Loading';
import { UserGreengrassCoreDevicesHook } from '../../hooks/GreengrassCoreDeviceHook';
import { requestApi } from '../../util/apis';
import {
  CreatedBy,
  FormControlElement,
  GreengrassCoreDeviceControl,
  GreengrassCoreDeviceDomIds,
  GreengrassCoreDevicePostResponse,
  MessageModalType
} from '../../util/types';
import { getErrorMessage } from '../../util/utils';

const logger = new Logger('GreengrassCoreDeviceForm');
const domIds: GreengrassCoreDeviceDomIds = {
  category: 'category',
  greengrassCoreDeviceName: 'greengrassCoreDeviceName'
};

/**
 * Renders the Greengrass core device registration form.
 * @returns The Greengrass core device modal
 */
export default function GreengrassCoreDeviceForm(): React.JSX.Element {
  const [confirmMessage, setConfirmMessage] = useState<string>('');
  const [createdBy, setCreatedBy] = useState<CreatedBy>(CreatedBy.SYSTEM);
  const [greengrassCoreDeviceName, setGreengrassCoreDeviceName] = useState<string>('');
  const [greengrassCoreDeviceStatus, setGreengrassCoreDeviceStatus] = useState<string>('');
  const [hasConfirmActionError, setHasConfirmActionError] = useState<boolean>(false);
  const [isGreengrassCoreDeviceNameValid, setIsGreengrassCoreDeviceNameValid] = useState<boolean>(false);
  const [loading, setLoading] = useState<boolean>(false);
  const [message, setMessage] = useState<string | React.ReactNode>('');
  const [showConfirmMessageModal, setShowConfirmMessageModal] = useState<boolean>(false);
  const [showMessageModal, setShowMessageModal] = useState<boolean>(false);
  const { close, greengrassCoreDevices, statusMap } = UserGreengrassCoreDevicesHook({
    setHasConfirmActionError,
    setMessage,
    setShowMessageModal
  });
  const change = (event: React.ChangeEvent<FormControlElement>) => {
    handleValueChange({
      domIds,
      event,
      greengrassCoreDevices,
      setCreatedBy,
      setGreengrassCoreDeviceName,
      setGreengrassCoreDeviceStatus,
      setIsGreengrassCoreDeviceNameValid,
      statusMap
    });
  };

  /**
   * Handles the Greengrass core device registration. It shows the confirmation modal.
   * @param event The Greengrass core device handle form event
   */
  function handleGreengrassCoreDevice(event: FormEvent): void {
    event.preventDefault();
    setConfirmMessage(I18n.get('info.message.register.greengrass.core.device').replace('{}', greengrassCoreDeviceName));

    if (createdBy === CreatedBy.USER && greengrassCoreDeviceStatus !== 'HEALTHY') {
      setConfirmMessage(
        I18n.get('warning.message.unhealthy.greengrass.core.device').replace('{}', greengrassCoreDeviceName)
      );
    }

    setShowConfirmMessageModal(true);
  }

  /**
   * Registers a Greengrass core device.
   */
  async function registerGreengrassCoreDevice(): Promise<void> {
    setLoading(true);

    try {
      const response = (await requestApi({
        method: 'post',
        path: '/greengrass',
        options: {
          body: {
            name: greengrassCoreDeviceName,
            control: GreengrassCoreDeviceControl.CREATE,
            createdBy
          }
        }
      })) as GreengrassCoreDevicePostResponse;

      setMessage(<span>{response.message}</span>);
      setHasConfirmActionError(false);
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      logger.error(errorMessage);

      setShowConfirmMessageModal(false);
      setHasConfirmActionError(true);
      setMessage(
        <Alert variant="danger">
          {I18n.get('error.message.register.greengrass.core.device')}
          <br />
          {I18n.get('error')}: <code>{JSON.stringify(errorMessage)}</code>
          <br />
        </Alert>
      );
    }

    setLoading(false);
    setShowMessageModal(true);
  }

  return (
    <Container>
      <Breadcrumb>
        <LinkContainer to="/greengrass">
          <Breadcrumb.Item>{I18n.get('manage.greengrass.core.devices')}</Breadcrumb.Item>
        </LinkContainer>
        <Breadcrumb.Item active>{I18n.get('register.greengrass.core.device')}</Breadcrumb.Item>
      </Breadcrumb>
      <Card>
        <Card.Body>
          <Form onSubmit={handleGreengrassCoreDevice}>
            <Form.Group>
              <Form.Label>
                {I18n.get('category')} <span className="red-text">*</span>
              </Form.Label>
              <Form.Text muted>{I18n.get('description.category')}</Form.Text>
              <Form.Control
                data-testid="category-select"
                id={domIds.category}
                as="select"
                onChange={change}
                value={createdBy}>
                <option value={CreatedBy.SYSTEM}>{I18n.get('category.create.new.greengrass.core.device')}</option>
                <option value={CreatedBy.USER}>{I18n.get('category.bring.existing.greengrass.core.device')}</option>
              </Form.Control>
            </Form.Group>
            <GreengrassCoreDeviceInputForm
              createdBy={createdBy}
              domIds={domIds}
              greengrassCoreDeviceName={greengrassCoreDeviceName}
              greengrassCoreDevices={greengrassCoreDevices}
              handleValueChange={change}
              isGreengrassCoreDeviceNameValid={isGreengrassCoreDeviceNameValid}
            />
            <EmptyRow />
            <Row>
              <Col className="justify-content-center grid">
                <ButtonGroup>
                  <Button
                    id="register-greengrass-core-device"
                    variant="primary"
                    size="sm"
                    type="submit"
                    disabled={
                      loading ||
                      (createdBy === CreatedBy.SYSTEM && !isGreengrassCoreDeviceNameValid) ||
                      (createdBy === CreatedBy.USER && greengrassCoreDevices.length === 0)
                    }>
                    <LoadingSpinner loading={loading} />
                    {I18n.get('register')}
                  </Button>
                  <EmptyCol />
                  <Button
                    id="register-greengrass-core-device"
                    variant="secondary"
                    size="sm"
                    onClick={() => close()}
                    disabled={loading}>
                    {I18n.get('cancel')}
                  </Button>
                </ButtonGroup>
              </Col>
            </Row>
          </Form>
        </Card.Body>
      </Card>
      <MessageModal
        show={showConfirmMessageModal}
        hide={() => setShowConfirmMessageModal(false)}
        message={confirmMessage}
        modalType={MessageModalType.CONFIRM}
        confirmAction={() => registerGreengrassCoreDevice()}
      />
      <MessageModal
        show={showMessageModal}
        hide={() => setShowMessageModal(false)}
        message={message}
        modalType={MessageModalType.MESSAGE}
        confirmAction={hasConfirmActionError ? () => undefined : () => close()}
      />
    </Container>
  );
}
