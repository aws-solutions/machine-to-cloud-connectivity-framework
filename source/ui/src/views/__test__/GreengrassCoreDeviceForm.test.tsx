// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { API } from '@aws-amplify/api';
import { I18n } from '@aws-amplify/core';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import GreengrassCoreDeviceForm from '../greengrass/GreengrassCoreDeviceForm';
import { CreatedBy, GreengrassCoreDeviceControl, OsPlatform, UserGreengrassCoreDeviceResponse } from '../../util/types';
import { API_NAME } from '../../util/utils';

const mockAPI = {
  get: jest.fn(),
  post: jest.fn()
};
API.get = mockAPI.get;
API.post = mockAPI.post;

const greengrassCoreDeviceName = 'mock-greengrass-core-device';
const userGreengrassCoreDeviceResponse: UserGreengrassCoreDeviceResponse = {
  greengrassCoreDevices: [
    { coreDeviceThingName: 'mock-greengrass-1', status: 'HEALTHY' },
    { coreDeviceThingName: 'mock-greengrass-2', status: 'UNHEALTHY' },
    { coreDeviceThingName: 'mock-greengrass-3', status: 'HEALTHY' }
  ]
};

beforeEach(() => {
  mockAPI.get.mockReset();
  mockAPI.post.mockReset();
});

test('renders the Greengrass core device form by default', async () => {
  mockAPI.get.mockResolvedValueOnce({ greengrassCoreDevices: [] });
  const greengrassCoreDeviceForm = render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );
  expect(greengrassCoreDeviceForm.container).toMatchSnapshot();
  expect(mockAPI.get).toHaveBeenCalledTimes(1);
  expect(mockAPI.get).toHaveBeenCalledWith(API_NAME, '/greengrass/user', {});
});

test('renders the Greengrass core device form by default with API failure', async () => {
  mockAPI.get.mockRejectedValueOnce({ message: 'GET failure' });
  render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );

  expect(await screen.findByText(/error.message.get.greengrass.core.devices.user/)).not.toBeNull();
  expect(mockAPI.get).toHaveBeenCalledTimes(1);
  expect(mockAPI.get).toHaveBeenCalledWith(API_NAME, '/greengrass/user', {});
});

test('tests cancel button', async () => {
  mockAPI.get.mockResolvedValueOnce({ greengrassCoreDevices: [] });
  render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass" />
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );

  expect(screen.queryByText(I18n.get('cancel'))).not.toBeNull();

  userEvent.click(screen.getByText(I18n.get('cancel')));
  expect(screen.queryByText(I18n.get('cancel'))).toBeNull();
});

test('tests handleValueChange function - category change when user Greengrass core devices are empty', async () => {
  mockAPI.get.mockResolvedValueOnce({ greengrassCoreDevices: [] });
  const { getByTestId, getByText } = render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );

  userEvent.selectOptions(getByTestId('category-select'), CreatedBy.USER);
  expect(
    (getByText(I18n.get('category.bring.existing.greengrass.core.device')) as HTMLOptionElement).selected
  ).toBeTruthy();
  expect(await screen.findByText(/no.available.greengrass.core.devices/)).not.toBeNull();
});

test('tests handleValueChange function - category change when user Greengrass core devices are not empty', async () => {
  mockAPI.get.mockResolvedValueOnce(userGreengrassCoreDeviceResponse);
  const { getByTestId, getByText } = render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );

  userEvent.selectOptions(getByTestId('category-select'), CreatedBy.USER);
  expect(
    (getByText(I18n.get('category.bring.existing.greengrass.core.device')) as HTMLOptionElement).selected
  ).toBeTruthy();

  const mockDevice = userGreengrassCoreDeviceResponse.greengrassCoreDevices[0];
  expect(await screen.findByText(`${mockDevice.coreDeviceThingName} (${mockDevice.status})`)).not.toBeNull();
}, 10000);

test('tests handleValueChange function - Greengrass core device name change with invalid value', async () => {
  mockAPI.get.mockResolvedValueOnce({ greengrassCoreDevices: [] });
  const { getByPlaceholderText } = render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );
  const input = getByPlaceholderText(I18n.get('placeholder.greengrass.core.device.name'));

  userEvent.type(input, '@@');
  expect((document.getElementById('greengrassCoreDeviceName') as HTMLInputElement).value).toEqual('@@');
  expect(await screen.findByText(/invalid.greengrass.core.device.name/)).not.toBeNull();
});

test('tests handleValueChange function - Greengrass core device name change with valid value', async () => {
  mockAPI.get.mockResolvedValueOnce({ greengrassCoreDevices: [] });
  const { getByPlaceholderText } = render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );
  const input = getByPlaceholderText(I18n.get('placeholder.greengrass.core.device.name'));

  userEvent.type(input, greengrassCoreDeviceName);
  expect((document.getElementById('greengrassCoreDeviceName') as HTMLInputElement).value).toEqual(
    greengrassCoreDeviceName
  );
});

test('tests handleGreengrassCoreDevice function - System', async () => {
  mockAPI.get.mockResolvedValueOnce({ greengrassCoreDevices: [] });
  const { getByPlaceholderText } = render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );
  const input = getByPlaceholderText(I18n.get('placeholder.greengrass.core.device.name'));

  userEvent.type(input, greengrassCoreDeviceName);
  expect((document.getElementById('greengrassCoreDeviceName') as HTMLInputElement).value).toEqual(
    greengrassCoreDeviceName
  );

  userEvent.click(screen.getByText(I18n.get('register')));
  expect(await screen.findByText(/info.message.register.greengrass.core.device/)).not.toBeNull();
}, 10000);

test('tests handleGreengrassCoreDevice function - User with HEALTHY device', async () => {
  mockAPI.get.mockResolvedValueOnce(userGreengrassCoreDeviceResponse);
  const { getByTestId, getByText } = render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );

  userEvent.selectOptions(getByTestId('category-select'), CreatedBy.USER);
  expect(
    (getByText(I18n.get('category.bring.existing.greengrass.core.device')) as HTMLOptionElement).selected
  ).toBeTruthy();

  const mockDevice = userGreengrassCoreDeviceResponse.greengrassCoreDevices[0];
  expect(await screen.findByText(`${mockDevice.coreDeviceThingName} (${mockDevice.status})`)).not.toBeNull();

  userEvent.selectOptions(getByTestId('user-greengrass-core-device-select'), mockDevice.coreDeviceThingName);
  expect(
    (getByText(`${mockDevice.coreDeviceThingName} (${mockDevice.status})`) as HTMLOptionElement).selected
  ).toBeTruthy();

  userEvent.click(screen.getByText(I18n.get('register')));
  expect(await screen.findByText(/info.message.register.greengrass.core.device/)).not.toBeNull();
});

test('tests handleGreengrassCoreDevice function - User with UNHEALTHY device', async () => {
  mockAPI.get.mockResolvedValueOnce(userGreengrassCoreDeviceResponse);
  const { getByTestId, getByText } = render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );

  userEvent.selectOptions(getByTestId('category-select'), CreatedBy.USER);
  expect(
    (getByText(I18n.get('category.bring.existing.greengrass.core.device')) as HTMLOptionElement).selected
  ).toBeTruthy();

  const mockDevice = userGreengrassCoreDeviceResponse.greengrassCoreDevices[1];
  expect(await screen.findByText(`${mockDevice.coreDeviceThingName} (${mockDevice.status})`)).not.toBeNull();

  userEvent.selectOptions(getByTestId('user-greengrass-core-device-select'), mockDevice.coreDeviceThingName);
  expect(
    (getByText(`${mockDevice.coreDeviceThingName} (${mockDevice.status})`) as HTMLOptionElement).selected
  ).toBeTruthy();

  userEvent.click(screen.getByText(I18n.get('register')));
  expect(await screen.findByText(/warning.message.unhealthy.greengrass.core.device/)).not.toBeNull();
});

test('tests registerGreengrassCoreDevice function - System success', async () => {
  mockAPI.get.mockResolvedValueOnce({ greengrassCoreDevices: [] });
  mockAPI.post.mockResolvedValueOnce({ message: 'Success' });
  const { getByPlaceholderText } = render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );
  const input = getByPlaceholderText(I18n.get('placeholder.greengrass.core.device.name'));

  userEvent.type(input, greengrassCoreDeviceName);
  expect((document.getElementById('greengrassCoreDeviceName') as HTMLInputElement).value).toEqual(
    greengrassCoreDeviceName
  );

  userEvent.click(screen.getByText(I18n.get('register')));
  expect(await screen.findByText(/info.message.register.greengrass.core.device/)).not.toBeNull();

  userEvent.click(screen.getByText(I18n.get('confirm')));
  expect(await screen.findByText('Success')).not.toBeNull();

  expect(mockAPI.post).toHaveBeenCalledTimes(1);
  expect(mockAPI.post).toHaveBeenCalledWith(API_NAME, '/greengrass', {
    body: {
      name: greengrassCoreDeviceName,
      control: GreengrassCoreDeviceControl.CREATE,
      createdBy: CreatedBy.SYSTEM,
      osPlatform: OsPlatform.LINUX
    }
  });
});

test('tests registerGreengrassCoreDevice function - User failure', async () => {
  mockAPI.get.mockResolvedValueOnce(userGreengrassCoreDeviceResponse);
  mockAPI.post.mockRejectedValueOnce({ errorMessage: 'Failure' });
  const { getByTestId, getByText } = render(
    <MemoryRouter initialEntries={['/greengrass/register']}>
      <Routes>
        <Route path="/greengrass/register" element={<GreengrassCoreDeviceForm />} />
      </Routes>
    </MemoryRouter>
  );

  userEvent.selectOptions(getByTestId('category-select'), CreatedBy.USER);
  expect(
    (getByText(I18n.get('category.bring.existing.greengrass.core.device')) as HTMLOptionElement).selected
  ).toBeTruthy();

  const mockDevice = userGreengrassCoreDeviceResponse.greengrassCoreDevices[0];
  expect(await screen.findByText(`${mockDevice.coreDeviceThingName} (${mockDevice.status})`)).not.toBeNull();

  userEvent.selectOptions(getByTestId('user-greengrass-core-device-select'), mockDevice.coreDeviceThingName);
  expect(
    (getByText(`${mockDevice.coreDeviceThingName} (${mockDevice.status})`) as HTMLOptionElement).selected
  ).toBeTruthy();

  userEvent.click(screen.getByText(I18n.get('register')));
  expect(await screen.findByText(/info.message.register.greengrass.core.device/)).not.toBeNull();

  userEvent.click(screen.getByText(I18n.get('confirm')));
  expect(await screen.findByText(/error.message.register.greengrass.core.device/)).not.toBeNull();

  expect(mockAPI.post).toHaveBeenCalledTimes(1);
  expect(mockAPI.post).toHaveBeenCalledWith(API_NAME, '/greengrass', {
    body: {
      name: mockDevice.coreDeviceThingName,
      control: GreengrassCoreDeviceControl.CREATE,
      createdBy: CreatedBy.USER,
      osPlatform: OsPlatform.LINUX
    }
  });
});
