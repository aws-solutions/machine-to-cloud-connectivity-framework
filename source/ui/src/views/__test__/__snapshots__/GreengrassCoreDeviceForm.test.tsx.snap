// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Greengrass core device form by default 1`] = `
<div>
  <div
    class="container"
  >
    <nav
      aria-label="breadcrumb"
    >
      <ol
        class="breadcrumb"
      >
        <li
          class="breadcrumb-item"
        >
          <a
            href="/greengrass"
          >
            manage.greengrass.core.devices
          </a>
        </li>
        <li
          aria-current="page"
          class="breadcrumb-item active"
        >
          register.greengrass.core.device
        </li>
      </ol>
    </nav>
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
        >
          <div
            class="form-group"
          >
            <label
              class="form-label"
            >
              category
               
              <span
                class="red-text"
              >
                *
              </span>
            </label>
            <small
              class="form-text text-muted"
            >
              description.category
            </small>
            <select
              class="form-control"
              data-testid="category-select"
              id="category"
            >
              <option
                value="System"
              >
                category.create.new.greengrass.core.device
              </option>
              <option
                value="User"
              >
                category.bring.existing.greengrass.core.device
              </option>
            </select>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
            >
              greengrass.core.device.name
               
              <span
                class="red-text"
              >
                *
              </span>
            </label>
            <small
              class="form-text text-muted"
            >
              description.greengrass.core.device.name
            </small>
            <input
              class="form-control is-invalid"
              id="greengrassCoreDeviceName"
              placeholder="placeholder.greengrass.core.device.name"
              required=""
              type="text"
              value=""
            />
            <div
              class="invalid-feedback"
            >
              invalid.greengrass.core.device.name
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <span>
                 
              </span>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="justify-content-left grid col"
            >
              <label
                class="form-label"
              >
                os.platform
                 
                <span
                  class="red-text"
                >
                  *
                </span>
              </label>
              <div>
                <div
                  class="form-check"
                >
                  <input
                    checked=""
                    class="form-check-input"
                    id="osPlatform"
                    type="radio"
                    value="linux"
                  />
                  <label
                    class="form-check-label"
                    for="osPlatform"
                    title=""
                  >
                    Linux
                  </label>
                </div>
                <div
                  class="form-check"
                >
                  <input
                    class="form-check-input"
                    id="osPlatform"
                    type="radio"
                    value="windows"
                  />
                  <label
                    class="form-check-label"
                    for="osPlatform"
                    title=""
                  >
                    Windows
                  </label>
                </div>
              </div>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <span>
                 
              </span>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="justify-content-center grid col"
            >
              <div
                class="btn-group"
                role="group"
              >
                <button
                  class="btn btn-primary btn-sm"
                  disabled=""
                  id="register-greengrass-core-device"
                  type="submit"
                >
                  register
                </button>
                <span>
                   
                </span>
                <button
                  class="btn btn-secondary btn-sm"
                  id="register-greengrass-core-device"
                  type="button"
                >
                  cancel
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
